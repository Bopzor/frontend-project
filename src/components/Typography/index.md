`<Typography />` component wraps a collection of symbols for quickly mocking up web applications, created by your friends at GoJob and conceived to be used in the FoG project.

Besides the somewhat obvious need they fill (people sometimes want to put headlines on top of things), the headline tags also have SEO value, help you to be a more organized writer, and make pages more user-friendly.

Read more: https://html.com/tags/heading/#ixzz5Jo977u2P

On a page with a single piece of content (which should be most pages), the main title for that content should be an `<Typography variant="main-title" />` tag. 

```js
  <Typography variant="main-title">h1 - This is a main title</Typography>
```

The main title could sometimes use a little subtitle to adjust the context of the message - for this - please use `<Typography variant="headline2" />` tag.

```js
  <Typography variant="headline2">h2 - This is a headline</Typography>
```

Major sections within that content should be headlined with `<Typography variant="headline3" />`.

```js
  <Typography variant="headline3">h3 - This is a section title</Typography>
```

Subsections within those sections should be headed with `<Typography variant="headline4" />`

```js
  <Typography variant="headline4">h4 - This is a section subtitle</Typography>
```

And this is the base text, that you can dive into! Just read this beautiful __lorem ipsum__ bellow!

To make one yourself, you can use the default `<Typography />` component.

```js
  <Typography>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Cras dapibus vulputate diam eu pretium. Mauris elit orci, ultricies id fermentum vel, porta et eros. Vestibulum condimentum lectus in convallis feugiat. Sed vulputate fringilla felis. Aliquam ut arcu et dui feugiat scelerisque eu quis diam. Mauris placerat congue dui sit amet blandit. Phasellus condimentum libero vel velit auctor, sit amet tincidunt velit varius.
    <br/><br/>
    Mauris lacinia porta faucibus. Fusce eu est ac eros vulputate mollis in ac felis. Aenean commodo scelerisque mi sed imperdiet. Donec at hendrerit nisi, eget vestibulum nisi. Sed sit amet magna luctus, facilisis erat quis, sagittis ligula. Aenean dignissim velit quis leo consequat ultricies. Proin quis pretium justo. Vestibulum at eros nisl. Fusce lobortis erat ante, eu cursus sapien molestie at. Pellentesque placerat ante diam, et euismod lacus dictum vel. Phasellus vitae sollicitudin mi.
    <br/><br/>
    In pulvinar eleifend convallis. Suspendisse elit erat, venenatis eget ullamcorper ut, laoreet iaculis nisl. Sed porta, felis id rhoncus aliquet, quam ipsum pellentesque metus, in sodales quam nunc vitae risus.
  </Typography>
```

You can render a label (generally rendered above inputs) with `<Typography variant="label" />`.

```js
  <Typography variant="label">Input headline</Typography>
```

Here are some other, more rare `<Typography>` variants:

```js
  const s = { style: { padding: 14 } };
  <div style={{ display: 'flex', flexDirection: 'column', placeItems: 'center' }}>
    <Typography variant="placeholder" {...s}>Placeholder</Typography>
    <Typography variant="table-item" {...s}>Table Item</Typography>
    <Typography variant="table-value" {...s}>Table Value</Typography>
    <Typography variant="link" {...s}>Text Link</Typography>
    <Typography variant="link-cancel" {...s}>Text Link - Cancel</Typography>
    <Typography variant="caption" {...s}>Caption</Typography>
    <Typography variant="breadcrumb" {...s}>Breadcrumb navigation</Typography>
  </div>
```

You can also apply a color to a text to indicate some states, such as __"text"__, __"primary"__, __"placeholder"__, __"alert"__, __"success"__ or __"warning"__.

```js
  <div style={{ display: 'flex', justifyContent: 'space-around' }}>
    <Typography color="text">I am text</Typography>
    <Typography color="primary">I am primary</Typography>
    <Typography color="placeholder">I am placeholder</Typography>
    <Typography color="alert">I am alert</Typography>
    <Typography color="success">I am success</Typography>
    <Typography color="warning">I am warning</Typography>
  </div>
```