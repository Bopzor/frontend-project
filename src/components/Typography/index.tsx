import * as React from 'react';

import { hot } from 'react-hot-loader';

import { css, cx } from 'emotion';
import memoizeOne from 'memoize-one';

import { Consumer, Theme } from '../../theme';

/**
 * Defines Typography variants
 */
export type TypographyVariant =
  | 'main-title'
  | 'headline2'
  | 'headline3'
  | 'headline4'
  | 'default'
  | 'label'
  | 'placeholder'
  | 'table-item'
  | 'table-value'
  | 'link'
  | 'link-cancel'
  | 'caption'
  | 'breadcrumb';

/**
 * Map to adjust element for each variant of Typography
 */
const variantMap: { [P in TypographyVariant]: keyof React.ReactHTML } = {
  'main-title': 'h1',
  headline2: 'h2',
  headline3: 'h3',
  headline4: 'h4',
  default: 'p',
  label: 'label',
  placeholder: 'span',
  'table-item': 'span',
  'table-value': 'span',
  link: 'a',
  'link-cancel': 'a',
  caption: 'span',
  breadcrumb: 'span'
};

/**
 * Typography component props
 */
export type TypographyProps = {
  /**
   * Chooses a variant of Typography
   */
  variant?: TypographyVariant;
  /**
   * Maps the color from theme
   */
  color?: keyof Theme['colors'];
  /**
   * Shorthand for text-align css prop
   */
  align?: React.CSSProperties['textAlign'];
  /**
   * Shorthand for flex css prop
   * @todo obsolete?
   */
  flex?: number;
  /**
   * Shorthand for text-transform css prop
   */
  transform?: React.CSSProperties['textTransform'];
} & (
  | React.HTMLProps<HTMLHeadingElement>
  | React.HTMLProps<HTMLParagraphElement>
  | React.HTMLProps<HTMLAnchorElement>
  | React.HTMLProps<HTMLSpanElement>
  | React.HTMLProps<HTMLLabelElement>);

class Typography extends React.Component<TypographyProps> {
  static defaultProps: Partial<TypographyProps> = {
    variant: 'default'
  };

  /**
   * Allows to generate an appropriate color classname
   * @param theme Theme
   * @param color color
   */
  getColorClass(theme: Theme, color: TypographyProps['color']) {
    return colorStyle(theme, color);
  }

  /**
   * Allows to choose an appropiate classname without disrupting memoization
   * @param theme Theme
   */
  getVariantStyle(theme: Theme) {
    switch (this.props.variant) {
      case 'main-title':
        return mainTitleStyle(theme);
      case 'headline2':
        return headline2Style(theme);
      case 'headline3':
        return headline3Style(theme);
      case 'headline4':
        return headline4Style(theme);
      case 'label':
        return labelStyle(theme);
      case 'placeholder':
        return placeholderStyle(theme);
      case 'table-item':
        return tableItemStyle(theme);
      case 'table-value':
        return tableValueStyle(theme);
      case 'link':
        return linkStyle(theme);
      case 'link-cancel':
        return linkCancelStyle(theme);
      case 'caption':
        return captionStyle(theme);
      case 'breadcrumb':
        return breadcrumbStyle(theme);
      default:
        return defaultStyle(theme);
    }
  }

  render() {
    const { variant, align, color, className, flex, transform, ...props } = this.props;
    return (
      <Consumer>
        {(theme) => {
          const classNames = cx(
            this.getVariantStyle(theme),
            this.getColorClass(theme, color),
            transform && transformStyle(transform),
            align && alignStyle(align),
            typeof flex !== 'undefined' && flexStyle(flex),
            className
          );
          return React.createElement(variantMap[variant!], { className: classNames, ...props });
        }}
      </Consumer>
    );
  }
}

export default hot(module)(Typography);

const colorStyle = memoizeOne(({ colors }: Theme, color: TypographyProps['color']) => {
  if (!color) {
    return '';
  }
  return css`
    color: ${colors[color]};
  `;
});
const transformStyle = memoizeOne((transform: TypographyProps['transform']) => {
  return css`
    text-transform: ${transform};
  `;
});
const flexStyle = memoizeOne((flex: TypographyProps['flex']) => {
  return css`
    flex: ${flex};
  `;
});
const alignStyle = memoizeOne((align: TypographyProps['align']) => {
  return css`
    text-align: ${align};
  `;
});
const mainTitleStyle = memoizeOne(({ colors }: Theme) => {
  return css`
    font-size: 28px;
    font-weight: 700;
    line-height: 1.21;
    letter-spacing: -0.1px;
    color: ${colors.text};
  `;
});
const headline2Style = memoizeOne(({ colors }: Theme) => {
  return css`
    font-size: 24px;
    font-weight: 600;
    letter-spacing: 0.3px;
    color: ${colors.text};
  `;
});
const headline3Style = memoizeOne(({ colors }: Theme) => {
  return css`
    font-size: 20px;
    font-weight: 600;
    letter-spacing: 0.3px;
    color: ${colors.text};
  `;
});
const headline4Style = memoizeOne(({ colors }: Theme) => {
  return css`
    font-size: 17px;
    font-weight: 600;
    letter-spacing: -0.2px;
    color: ${colors.text};
  `;
});
const defaultStyle = memoizeOne(({ colors }: Theme) => {
  return css`
    font-size: 15px;
    font-weight: 400;
    line-height: 1.33;
    letter-spacing: -0.2px;
    color: ${colors.text};
  `;
});
const labelStyle = memoizeOne(({ colors }: Theme) => {
  return css`
    font-size: 12px;
    font-weight: 500;
    line-height: 2.08;
    letter-spacing: 0.2px;
    text-transform: uppercase;
    color: ${colors.placeholder};
  `;
});
const placeholderStyle = memoizeOne(({ colors }: Theme) => {
  return css`
    font-size: 15px;
    line-height: 1.33;
    color: ${colors.placeholder};
  `;
});
const tableItemStyle = memoizeOne(({ colors }: Theme) => {
  return css`
    font-size: 15px;
    font-weight: 500;
    line-height: 1.6;
    color: ${colors.text};
  `;
});
const tableValueStyle = memoizeOne(({ colors }: Theme) => {
  return css`
    font-size: 15px;
    font-weight: 400;
    line-height: 1.33;
    color: ${colors.text};
  `;
});
const linkStyle = memoizeOne(({ colors }: Theme) => {
  return css`
    font-size: 15px;
    font-weight: 400;
    line-height: 1.33;
    letter-spacing: -0.2px;
    cursor: pointer;
    color: ${colors.primary};
    :hover {
      text-decoration: underline;
    }
  `;
});
const linkCancelStyle = memoizeOne(({ colors }: Theme) => {
  return css`
    font-size: 15px;
    font-weight: 400;
    line-height: 1.33;
    letter-spacing: -0.2px;
    cursor: pointer;
    color: ${colors.placeholder};
    :hover {
      text-decoration: underline;
    }
  `;
});
const captionStyle = memoizeOne(({ colors }: Theme) => {
  return css`
    font-size: 12px;
    font-weight: 400;
    line-height: 1.67;
    letter-spacing: -0.2px;
    color: ${colors.placeholder};
  `;
});
const breadcrumbStyle = memoizeOne(({ colors }: Theme) => {
  return css`
    font-size: 14px;
    font-weight: 400;
    color: ${colors.placeholder};
  `;
});
