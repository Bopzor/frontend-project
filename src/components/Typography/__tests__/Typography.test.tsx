import * as React from 'react';

import { render } from 'enzyme';
import { renderToJson } from 'enzyme-to-json';

import Typography from '../';

describe('Button', () => {
  describe('Should match snapshot for', () => {
    test('default', () => {
      const cheerio = render(<Typography>Hello</Typography>);
      expect(renderToJson(cheerio)).toMatchSnapshot();
    });
    test('disabled', () => {
      const cheerio = render(<Typography disabled>Hello</Typography>);
      expect(renderToJson(cheerio)).toMatchSnapshot();
    });
    test('with variants', () => {
      const cheerio = render(
        <div>
          <Typography variant="placeholder">Placeholder</Typography>
          <Typography variant="table-item">Table Item</Typography>
          <Typography variant="table-value">Table Value</Typography>
          <Typography variant="link">Text Link</Typography>
          <Typography variant="link-cancel">Text Link - Cancel</Typography>
          <Typography variant="caption">Caption</Typography>
          <Typography variant="breadcrumb">Breadcrumb navigation</Typography>
        </div>
      );
      expect(renderToJson(cheerio)).toMatchSnapshot();
    });
    test('with colors', () => {
      const cheerio = render(
        <div>
          <Typography color="text">I am text</Typography>
          <Typography color="primary">I am primary</Typography>
          <Typography color="placeholder">I am placeholder</Typography>
          <Typography color="alert">I am alert</Typography>
          <Typography color="success">I am success</Typography>
          <Typography color="warning">I am warning</Typography>
        </div>
      );
      expect(renderToJson(cheerio)).toMatchSnapshot();
    });
  });
});
