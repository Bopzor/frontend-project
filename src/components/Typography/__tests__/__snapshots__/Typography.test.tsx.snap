// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Should match snapshot for default 1`] = `
.emotion-0 {
  font-size: 15px;
  font-weight: 400;
  line-height: 1.33;
  -webkit-letter-spacing: -0.2px;
  -moz-letter-spacing: -0.2px;
  -ms-letter-spacing: -0.2px;
  letter-spacing: -0.2px;
  color: #2a225b;
}

<p
  class="emotion-0"
>
  Hello
</p>
`;

exports[`Button Should match snapshot for disabled 1`] = `
.emotion-0 {
  font-size: 15px;
  font-weight: 400;
  line-height: 1.33;
  -webkit-letter-spacing: -0.2px;
  -moz-letter-spacing: -0.2px;
  -ms-letter-spacing: -0.2px;
  letter-spacing: -0.2px;
  color: #2a225b;
}

<p
  class="emotion-0"
  disabled=""
>
  Hello
</p>
`;

exports[`Button Should match snapshot for with colors 1`] = `
.emotion-0 {
  font-size: 15px;
  font-weight: 400;
  line-height: 1.33;
  -webkit-letter-spacing: -0.2px;
  -moz-letter-spacing: -0.2px;
  -ms-letter-spacing: -0.2px;
  letter-spacing: -0.2px;
  color: #2a225b;
  color: #2a225b;
}

.emotion-1 {
  font-size: 15px;
  font-weight: 400;
  line-height: 1.33;
  -webkit-letter-spacing: -0.2px;
  -moz-letter-spacing: -0.2px;
  -ms-letter-spacing: -0.2px;
  letter-spacing: -0.2px;
  color: #2a225b;
  color: #5856d6;
}

.emotion-2 {
  font-size: 15px;
  font-weight: 400;
  line-height: 1.33;
  -webkit-letter-spacing: -0.2px;
  -moz-letter-spacing: -0.2px;
  -ms-letter-spacing: -0.2px;
  letter-spacing: -0.2px;
  color: #2a225b;
  color: #9fa6c7;
}

.emotion-3 {
  font-size: 15px;
  font-weight: 400;
  line-height: 1.33;
  -webkit-letter-spacing: -0.2px;
  -moz-letter-spacing: -0.2px;
  -ms-letter-spacing: -0.2px;
  letter-spacing: -0.2px;
  color: #2a225b;
  color: #fd5573;
}

.emotion-4 {
  font-size: 15px;
  font-weight: 400;
  line-height: 1.33;
  -webkit-letter-spacing: -0.2px;
  -moz-letter-spacing: -0.2px;
  -ms-letter-spacing: -0.2px;
  letter-spacing: -0.2px;
  color: #2a225b;
  color: #24bdb9;
}

.emotion-5 {
  font-size: 15px;
  font-weight: 400;
  line-height: 1.33;
  -webkit-letter-spacing: -0.2px;
  -moz-letter-spacing: -0.2px;
  -ms-letter-spacing: -0.2px;
  letter-spacing: -0.2px;
  color: #2a225b;
  color: #fd9727;
}

<div>
  <p
    class="emotion-0"
  >
    I am text
  </p>
  <p
    class="emotion-1"
  >
    I am primary
  </p>
  <p
    class="emotion-2"
  >
    I am placeholder
  </p>
  <p
    class="emotion-3"
  >
    I am alert
  </p>
  <p
    class="emotion-4"
  >
    I am success
  </p>
  <p
    class="emotion-5"
  >
    I am warning
  </p>
</div>
`;

exports[`Button Should match snapshot for with variants 1`] = `
.emotion-0 {
  font-size: 15px;
  line-height: 1.33;
  color: #9fa6c7;
}

.emotion-1 {
  font-size: 15px;
  font-weight: 500;
  line-height: 1.6;
  color: #2a225b;
}

.emotion-2 {
  font-size: 15px;
  font-weight: 400;
  line-height: 1.33;
  color: #2a225b;
}

.emotion-3 {
  font-size: 15px;
  font-weight: 400;
  line-height: 1.33;
  -webkit-letter-spacing: -0.2px;
  -moz-letter-spacing: -0.2px;
  -ms-letter-spacing: -0.2px;
  letter-spacing: -0.2px;
  cursor: pointer;
  color: #5856d6;
}

.emotion-3:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-4 {
  font-size: 15px;
  font-weight: 400;
  line-height: 1.33;
  -webkit-letter-spacing: -0.2px;
  -moz-letter-spacing: -0.2px;
  -ms-letter-spacing: -0.2px;
  letter-spacing: -0.2px;
  cursor: pointer;
  color: #9fa6c7;
}

.emotion-4:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-5 {
  font-size: 12px;
  font-weight: 400;
  line-height: 1.67;
  -webkit-letter-spacing: -0.2px;
  -moz-letter-spacing: -0.2px;
  -ms-letter-spacing: -0.2px;
  letter-spacing: -0.2px;
  color: #9fa6c7;
}

.emotion-6 {
  font-size: 14px;
  font-weight: 400;
  color: #9fa6c7;
}

<div>
  <span
    class="emotion-0"
  >
    Placeholder
  </span>
  <span
    class="emotion-1"
  >
    Table Item
  </span>
  <span
    class="emotion-2"
  >
    Table Value
  </span>
  <a
    class="emotion-3"
  >
    Text Link
  </a>
  <a
    class="emotion-4"
  >
    Text Link - Cancel
  </a>
  <span
    class="emotion-5"
  >
    Caption
  </span>
  <span
    class="emotion-6"
  >
    Breadcrumb navigation
  </span>
</div>
`;
